### PR TITLE
New feature to link from endpoints to corresponding view source code.

### PR DESCRIPTION
**Pull request by kenrobbins**
Supports code in the docs with viewcode or generating a simple external link
with a prefix.

@sontek Thoughts?

### CODE REVIEW

**Comment by kenrobbins** [comment]
@sontek ping

**Comment by sontek** [comment]
@kenrobbins Sorry, I reviewed this over the weekend and the concept/implementation is good in my mind but wanted to pull it down and play with it to understand how it all goes together.

**Comment by kenrobbins** [comment]
Cool. No rush. Just wasn't sure if it was missed over the holiday.

File: README.rst [hunk 1]
```diff
@@ -72,4 +72,41 @@ there are a few options you can use:
         :match-path:
           ^/data
 
+Linking to Source Code
+----------------------
+
+If you want to link from the endpoint to the source code for the corresponding
+views and you are using sphinx.ext.viewcode_, you can generate links to the
```
Comment (sontek) [inline README.rst]: What happens if they use it and `sphinx.ext.viewcode` isn't available?  Should we enable it if they use this feature like we do httpdomain?

File: README.rst [hunk 2]
```diff
@@ -72,4 +72,41 @@ there are a few options you can use:
...
+source code pages it generates.  Alternatively, if your source is on the web,
+you can generate external links instead.
+
+- ``:link-code:`` - Enable links from endpoints to source code.  Assumes
+  sphinx.ext.viewcode_ is being used unless ``link-code-base`` is specified.
+- ``:link-code-base:`` - Base URL for generating links to source code.  Links
+  are of the format
+  ``{link-code-base}/{module-path}#L{lineno-start}-L{lineno-end}``.
```
Comment (sontek) [inline README.rst]: This format might be something we want to allow being customized.  Have we looked at popular sites like bitbucket, github, and gitlab to see if they all match this?

File: pyramid_autodoc/utils.py [hunk 3]
```diff
@@ -64,6 +65,38 @@ def _get_request_methods(route_request_methods, view_request_methods):
...
 
+def _get_view_source(view_callable):
+    if not view_callable or not hasattr(view_callable, '__name__'):
+        return {}
+
+    view_module_name = view_callable.__module__
+    try:
+        view_callable_name = '%s.%s' % (
+            view_callable.im_class.__name__,
+            view_callable.__name__)
```
Comment (sontek) [inline pyramid_autodoc/utils.py]: I'm not sure if this is going to be python3 compatible

File: README.rst (hunk 2 above)
Comment (kenrobbins) [inline README.rst]: That's a good point. I made this as a quick implementation, but I thought about using an implementation like `sphinx.ext.linkcode`, where the user specifies a callback function to create the links. Seems like that might be best if we want this feature. Maybe it's not worth it though.

File: README.rst (hunk 1 above)
Comment (kenrobbins) [inline README.rst]: Yeah, we can do that. If `sphinx.ext.viewcode` isn't available, we still create the `pending_xref`, but `viewcode` doesn't resolve the links. I kind of figured if you specify `:link-code:`, you'd be ready for it.

File: pyramid_autodoc/utils.py (hunk 3 above)
Comment (kenrobbins) [inline pyramid_autodoc/utils.py]: Oops.

File: README.rst (hunk 1 above)
Comment (kenrobbins) [inline README.rst]: How about this? https://github.com/SurveyMonkey/pyramid_autodoc/commit/dac1cbe29a164a2094630ea8fe979c817f471221#diff-46f5c6fd135691e50a9e9d06a00e399dR230

File: pyramid_autodoc/utils.py (hunk 3 above)
Comment (kenrobbins) [inline pyramid_autodoc/utils.py]: https://github.com/SurveyMonkey/pyramid_autodoc/commit/dac1cbe29a164a2094630ea8fe979c817f471221#diff-a0cb574de9cda02564ce68a106b21a0eR75

File: README.rst (hunk 2 above)
Comment (kenrobbins) [inline README.rst]: I updated to be a pattern instead. https://github.com/SurveyMonkey/pyramid_autodoc/commit/dac1cbe29a164a2094630ea8fe979c817f471221#diff-88b99bb28683bd5b7e3a204826ead112R108

I'm thinking this is better than before but not as heavy weight as asking people to implement a callback function.

**Pull request merged**